core/src/connection/manager.rs

```diff
@@ -0,0 +1,483 @@
+                Poll::Ready(Some(event)) => {
+                    if self.tasks.contains_key(event.id()) { // (1)
+                        break event
+                    }
+                }
+                Poll::Pending => return Poll::Pending,
+                Poll::Ready(None) => unreachable!("Manager holds both sender and receiver."),
+            }
+        };
+
+        if let hash_map::Entry::Occupied(mut task) = self.tasks.entry(*event.id()) {
+            Poll::Ready(match event {
+                task::Event::Notify { id: _, event } =>
+                    Event::ConnectionEvent {
+                        entry: EstablishedEntry { task },
+                        event
+                    },
+                task::Event::Established { id: _, info } => { // (2)
+                    task.get_mut().state = TaskState::Established(info); // (3)
+                    Event::ConnectionEstablished {
+                        entry: EstablishedEntry { task },
+                    }
+                }
+                task::Event::Error { id, error, handler } => {
+                    let id = ConnectionId(id);
+                    let task = task.remove();
+                    match handler {
+                        Some(handler) =>
+                            Event::PendingConnectionError { id, error, handler },
+                        None => match task.state {
+                            TaskState::Established(connected) =>
+                                Event::ConnectionError { id, connected, error },
+                            TaskState::Pending => unreachable!(
+                                "handler = `None` implies (2) occurred on that task and thus (3)."
+                            ),
+                        }
+                    }
+                }
+
+            })
+        } else {
+            unreachable!("By (1)")
+        }
+    }
+}
+
+/// An entry for a connection in the manager.
+#[derive(Debug)]
+pub enum Entry<'a, I, C> {
+    Pending(PendingEntry<'a, I, C>),
+    Established(EstablishedEntry<'a, I, C>)
+}
+
+impl<'a, I, C> Entry<'a, I, C> {
+    fn new(task: hash_map::OccupiedEntry<'a, TaskId, TaskInfo<I, C>>) -> Self {
+        match &task.get().state {
+            TaskState::Pending => Entry::Pending(PendingEntry { task }),
+            TaskState::Established(_) => Entry::Established(EstablishedEntry { task })
+        }
+    }
+}
+
+/// An entry for a managed connection that is considered established.
+#[derive(Debug)]
+pub struct EstablishedEntry<'a, I, C> {
+    task: hash_map::OccupiedEntry<'a, TaskId, TaskInfo<I, C>>,
+}
+
+impl<'a, I, C> EstablishedEntry<'a, I, C> {
+    /// (Asynchronously) sends an event to the connection handler.
+    ///
+    /// **Note:** Must only be called after `poll_ready_notify_handler`
+    /// was successful and without interference by another thread, otherwise the event is discarded.
+    pub fn notify_handler(&mut self, event: I) {
```

Comment:
Why not replacing this racy `poll_ready_notify_handler`/`notify_handler` combo with a non-racy version, e.g.

```rust
pub fn notify_handler(&mut self, event: I, cx: &mut Context) -> Option<I> {
    if self.poll_ready_notify_task(cx).is_pending() {
        return Some(event)
    }
    let cmd = task::Command::NotifyHandler(event);
    self.notify_task(cmd);
    None
}
```

Reply:
This would be similar to the old `AsyncSink` enum, and it forces you to potentially store the event somewhere for later when the handler is ready. Meanwhile, with this "racy" API (it's not racy, since it requires `&mut`), you can generate the event on-the-fly.

Reply:
We had this discussion in https://github.com/libp2p/rust-libp2p/pull/1373 already.

Reply:
My preference would be to start with the simpler and safer API as per the suggestion of @twittner and potentially add the `poll_ready_x` / `x` combination later as an alternative if needed. That is, we may at the end offer both, just like `futures::channel::mpsc::Sender` does in form of `try_send`, `start_send` and `poll_ready`.

To elaborate, I'm sure we all understand the motivation and trade-off behind these two styles of API and we just have different priorities w.r.t. what we consider more important. Given that the `poll_ready_x` / `x` design is arguably both more "complex" (in that you have to use 2 methods instead of 1) and more error-prone (in that you have to use these two methods in a certain way, subject as well to pitfalls w.r.t multi-threading as already discussed in https://github.com/libp2p/rust-libp2p/pull/1373) I personally prefer to start with just providing `try_notify_handler` for now. The `poll_ready_notify_handler` / `notify_handler` combination could be added later if there is a strong demand or use-case (which as far as I can tell even the `Swarm` does not currently have).

Does that make sense and sound reasonable?

Reply:
@tomaka Could you elaborate on the :-1: w.r.t. what aspect you disagree with? Do you insist on offering `poll_ready_notify_handler` / `notify_handler` from the start or do you even have objections to offering `try_notify_handler`?

Reply:
> it's not racy, since it requires `&mut`

This proposition has been refuted in #1373 and I am surprised to read it here again. I also do not understand why one would want to ignore the safer alternative to the current API. If one really wants to offer the more dangerous poll-then-send API I think @romanb's suggestion makes perfect sense.

Reply:
> This proposition has been refuted in #1373

It hasn't been refuted, I just gave up arguing.

> the safer alternative

The definition "safe" in the context of Rust is "cannot trigger undefined behaviour". In no way misusing this API can trigger any undefined behaviour. The word "safe" is totally irrelevant here.

The worst that can happen is a panic resulting in us misusing the API. There are lots of places in the Rust standard library and in the Rust ecosystem where misusing the API will result in a panic. Polling a `Future` after it has returned `Ready`, to give the most prominent example.

Using this API in a correct way is as racy as `try_notify_handler` is, in the sense that sometimes it will return `Pending` and sometimes it will return `Ready`, depending on the timing of the other tasks/threads.

Misuing this API is racy in a similar way as using this API correct, with the exception that the implementation might panic (because you misused the API).

> the more dangerous

In order for me to change my point of view, I would like to get a precise definition of what "dangerous" means here for you.

Reply:
My issue with `try_notify_handler` is the fact that there is no way for you to know whether your event will be accepted before generating said event.

An API that I think is a good is:

```
impl Manager {
    /// Try to give out a slot for sending an event.
    pub fn poll_ready_notify(&mut self, ...) -> Poll<EventSend> { ... }
}

impl<'a, T> EventSend<'a, T> {
    /// Sends the event, consuming `self`.
    pub fn send_event(self, event: T) { ... }
}
```

Reply:
> The worst that can happen is a panic resulting from us misusing the API. There are lots of places in the Rust standard library and in the Rust ecosystem where misusing the API will result in a panic. Polling a `Future` after it has returned `Ready`, to give the most prominent example.

Note though that the underlying API we're talking about here, `futures::channel::mpsc::Sender`, never panics (and neither does our current API). Indeed the `Sender` API says `poll_ready` _should_ be called before `start_send` and it consequently still returns an error if the channel is full. It is a bit telling, to me anyway, that they did not go with saying "`poll_ready` _must_ be called before `start_send`" combined with panicking due to programmer error on `start_send` if the channel is full. And they also provide `try_send`.

> 
> Using this API in a correct way is as racy as `try_notify_handler` is, in the sense that sometimes it will return `Pending` and sometimes it will return `Ready`, depending on the timing of the other tasks/threads.

Here we are again at the point which I thought we clarified e.g. [here](https://github.com/libp2p/rust-libp2p/pull/1373#issuecomment-571192286). To recap, the crucial difference is in the assumptions that the method that must be called second (e.g. `notify_handler`) places on the possible errors that can occur and how it deals with them. If it assumes certain errors can never happen _because of a requirement that another method must be called first_ then calls to such two functions which may be interleaved by multiple threads if e.g. `Mutex`es are used can lead to unexpected behaviour, hence why I already augmented the commentary on `notify_handler` to say that not only must `poll_ready_notify_handler` be called first, but also that no other thread must be allowed to interleave between `poll_ready_notify_handler` and `notify_handler`. Of course that is easy to satisfy if, as a user, you just call both these `&mut` methods in sequence on the same thread without e.g. going through a `Mutex`, but nevertheless it is another thing to watch out for. We may agree to disagree on whether that justifies calling an API "racy", but in any case the `futures::channel::mpsc::Sender` API is actually not racy, whatever definition you choose, whereas the API provided by us, currently, is considered racy by myself and @twittner because it may silently (well, now it logs at least) ignore errors due to the channel being full when calling `notify_handler`.

I have added `try_notify_handler` in a new commit in the hope that it is uncontroversial since it follows the example of the `futures::channel::mpsc::Sender` API, leaving the other functions in place. Can we all live with that?

Reply:
> > This proposition has been refuted in #1373
> 
> It hasn't been refuted, I just gave up arguing.

I would love to hear a new argument then. If "it's not racy, since it requires `&mut`" would be true why would we even have this discussion? Since the outcome of concurrent calls to `poll_ready_notify_handler` and `notify_handler` depends on relative timings there clearly is a race condition here that the individual `&mut`s in each of these methods can not prevent.

> 
> > the safer alternative
> 
> The definition of "safe" in the context of Rust is "cannot trigger undefined behaviour". In no way misusing this API can trigger any undefined behaviour. The word "safe" is totally irrelevant here.

[...]

> > the more dangerous
> 
> In order for me to change my point of view, I would like to get a precise definition of what "dangerous" means here for you.

Rust does not have a monopoly on the word "safe" and since you were asking for a definiton of "dangerous", I am happy to elaborate: Assuming there are undesirable program states, e.g. a panic, an API is safe w.r.t. those states if it can by construction guarantee that those states can never occur. Correspondingly an API is dangerous w.r.t. those states if it can not give those guarantees.

> Misuing this API is racy in a similar way as using this API correctly, with the exception that you might get a panic (because you misused the API).

This "exception" being an undesirable program state and the fact that the API can not guarantee its own correct use but allows for degrees of freedom that can lead to such a state makes it more dangerous in my book.

> My issue with `try_notify_handler` is the fact that there is no way for you to know whether your event will be accepted before generating said event.

We could combine both worlds by keeping `poll_ready_notify_handler` as an *indication* that *most likely* `notify_handler` will be able to accept the event, but in case it can not (because of API misuse) we can at least return the event:


```rust
pub fn poll_ready_notify_handler(&mut self, cx: &mut Context) -> Poll<()>;

pub fn notify_handler(&mut self, event: I, cx: &mut Context) -> Option<I>;
```

Reply:
> We could combine both worlds by keeping `poll_ready_notify_handler` as an _indication_ that _most likely_ `notify_handler` will be able to accept the event, but in case it can not (because of API misuse) we can at least return the event:
> 
> ```rust
> pub fn poll_ready_notify_handler(&mut self, cx: &mut Context) -> Poll<()>;
> 
> pub fn notify_handler(&mut self, event: I, cx: &mut Context) -> Option<I>;
> ```

I would be ok with that as well, and it would make `try_notify_handler` superfluous (`notify_handler` would then internally use `Sender::try_send` which is used  by `Sender::start_send` internally anyway).